File: 1110-Delete Nodes And Return Forest.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def delNodes(self, root, to_delete):
        """
        :type root: TreeNode
        :type to_delete: List[int]
        :rtype: List[TreeNode]
        """        
# first visited
        deleteSet = set(to_delete)
        self.res = []
        
        def dfs(node, parentExist):
            if not node:
                return 
            exist = node.val not in deleteSet
            if not parentExist and exist:
                self.res.append(node)
            dfs(node.left, exist)
            dfs(node.right, exist)
            if node.left and node.left.val in deleteSet:
                node.left = None
            if node.right and node.right.val in deleteSet:
                node.right = None
        dfs(root, False)
        return self.res
    
# simplified
        deleteSet = set(to_delete)
        self.res = []
        
        def dfs(node, parentExist):
            if not node:
                return 
            exist = node.val not in deleteSet
            if not parentExist and exist:
                self.res.append(node)
            node.left = dfs(node.left, exist)
            node.right = dfs(node.right, exist)
            return node if exist else None
        dfs(root, False)
        return self.res

File: JavaScript/1110-Delete Nodes And Return Forest.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number[]} to_delete
 * @return {TreeNode[]}
 */
var delNodes = function(root, to_delete) {
    
    const deleteSet = new Set(to_delete);
    let res = [];
    
    let dfs = (node, parentExist) =>{
        if (!node) return null;
        const exist = !deleteSet.has(node.val);
        if (exist && !parentExist){
            res.push(node);
        }
        node.left = dfs(node.left, exist);
        node.right = dfs(node.right, exist);
        
        return exist? node: null;
        
    }
    dfs(root, false);
    
    return res;
};